### PR TITLE
Enhancement animation loading

### PR DESCRIPTION
**Pull request by BenjaminDeleze**
After seeing the animation generated by Robocup (https://robocup-hlvs.de/GAME_K-GD6-3/public_logs/K-GD6-3.html) , I improved the loading time of the animation by:

- starting the loading of JSON ealier.
- Then I generated some keyframes (think about it like snapshots of the simulation) in order to increase the response time when jumping forward or backward in the replay. The keyframes are generated during the loading.
  - I measured an overhead of 300ms on the Robocup animation for a total loading time of about 16 seconds. It seems acceptable to me knowing that this animation has a JSON file of 400MB.
- I also reordered some loops to minimize the number of iterations
- Finally I made some cleanup.

To test and compare with the original: https://cyberbotics.com/animations/robocup/K-GD6-3.html

### CODE REVIEW

File: resources/web/wwi/Animation.js [hunk 1]
```diff
@@ -89,75 +157,123 @@ export default class Animation {
...
+          previousPoseStep = (closestKeyFrame + 1) * this.keyFrameStepSize;
+          if (this._keyFrames.size > 0)
+            previousStepIsAKeyFrame = true;
+        }
+
+        const completeIds = new Set();
+        const appliedFieldsByIds = new Map();
+        const appliedLabelsIds = new Set();
+
+        // We do not want to include the previousPoseStep in the loop as its updates are in the keyFrame. However, we need to include it if their is no keyFrames or if it is the step 0 as their is no keyFrame for it
```
Comment (omichel) [inline resources/web/wwi/Animation.js:178]: their -> there

File: resources/web/wwi/Animation.js [hunk 2]
```diff
@@ -89,75 +157,123 @@ export default class Animation {
...
+        if (previousStepIsAKeyFrame || previousPoseStep !== 0)
+          previousPoseStep++;
+
+        for (let i = this.step; i >= previousPoseStep; i--) { // Iterate through each step until the nearest keyFrame is reached or all necessary updates have been applied. Go in decreasing order to minize the number of step.
```
Comment (omichel) [inline resources/web/wwi/Animation.js:182]: steps

**Review by omichel — CHANGES_REQUESTED** [review]
That looks good to me.

**Comment by omichel** [comment]
It works very well! However, when testing on both Firefox and Chrome, there is a waiting time after the scene is display and before the buttons of the GUI are displayed. What is webotsjs doing at this point? It would be nice to display a progress message here, maybe, otherwise the user may believe that it is just displaying a static picture...

**Comment by BenjaminDeleze** [comment]
> It works very well! However, when testing on both Firefox and Chrome, there is a waiting time after the scene is displayed and before the buttons of the GUI are displayed. What is webotsjs doing at this point? It would be nice to display a progress message here, maybe, otherwise the user may believe that it is just displaying a static picture...

It is when the scene has finished to load, but not the json.
Indeed it would be nice, but as @ThomasOliverKimble is reworking the progress bar, I thought it was maybe better to wait until the new version is out.

File: resources/web/wwi/Animation.js [hunk 3]
```diff
@@ -89,75 +157,123 @@ export default class Animation {
...
+          previousPoseStep = (closestKeyFrame + 1) * this.keyFrameStepSize;
+          if (this._keyFrames.size > 0)
+            previousStepIsAKeyFrame = true;
+        }
+
+        const completeIds = new Set();
+        const appliedFieldsByIds = new Map();
+        const appliedLabelsIds = new Set();
+
+        // We do not want to include the previousPoseStep in the loop as its updates are in the keyFrame. However, we need to include it if there is no keyFrames or if it is the step 0 as their is no keyFrame for it
```
Comment (omichel) [inline resources/web/wwi/Animation.js:178]:
```suggestion
        // We do not want to include the previousPoseStep in the loop as its updates are in the keyFrame. However, we need to include it if there is no keyFrames or if it is the step 0 as there is no keyFrame for it
```

**Review by omichel — CHANGES_REQUESTED** [review]

**Comment by omichel** [comment]
Not sure why the mac source tests are failing...

File: resources/web/wwi/Animation.js (hunk 3 above)
Comment (BenjaminDeleze) [reply resources/web/wwi/Animation.js:178]: oups

**Review by omichel — APPROVED** [review]
Thank you.

**Pull request merged**